go-controller/pkg/clustermanager/network_cluster_controller_test.go

```diff
@@ -0,0 +1,203 @@
+	"github.com/onsi/gomega"
+	"github.com/urfave/cli/v2"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	ovntest "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/testing"
+	ovntypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+var _ = ginkgo.Describe("Network Cluster Controller", func() {
+	var (
+		app      *cli.App
+		f        *factory.WatchFactory
+		stopChan chan struct{}
+		wg       *sync.WaitGroup
+	)
+
+	ginkgo.BeforeEach(func() {
+		// Restore global default values before each testcase
+		gomega.Expect(config.PrepareTestConfig()).To(gomega.Succeed())
+
+		app = cli.NewApp()
+		app.Name = "test"
+		app.Flags = config.Flags
+		stopChan = make(chan struct{})
+		wg = &sync.WaitGroup{}
+	})
+
+	ginkgo.AfterEach(func() {
+		close(stopChan)
+		if f != nil {
+			f.Shutdown()
+		}
+		wg.Wait()
+	})
+
+	ginkgo.Context("Host Subnets", func() {
+		ginkgo.It("removes an unused dual-stack allocation from single-stack cluster", func() {
+			app.Action = func(ctx *cli.Context) error {
+				nodes := []v1.Node{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "node1",
+							Annotations: map[string]string{
+								"k8s.ovn.org/node-subnets": "{\"default\":[\"10.128.0.0/24\", \"fd02:0:0:2::2895/64\"]}",
+							},
+						},
+					},
+				}
+				kubeFakeClient := fake.NewSimpleClientset(&v1.NodeList{
+					Items: nodes,
+				})
+				fakeClient := &util.OVNClusterManagerClientset{
+					KubeClient: kubeFakeClient,
+				}
+
+				_, err := config.InitConfig(ctx, nil, nil)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				config.Kubernetes.HostNetworkNamespace = ""
+
+				f, err = factory.NewClusterManagerWatchFactory(fakeClient)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				err = f.Start()
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+				ncc := newNetworkClusterController(ovntypes.DefaultNetworkName, config.Default.ClusterSubnets,
+					fakeClient, f, false, &util.DefaultNetInfo{}, &util.DefaultNetConfInfo{})
+				ncc.Start(ctx.Context)
+				defer ncc.Stop()
+
+				// Check that the default network controller removes the unused v6 node subnet annotation
+				gomega.Eventually(func() ([]*net.IPNet, error) {
+					updatedNode, err := fakeClient.KubeClient.CoreV1().Nodes().Get(context.TODO(), nodes[0].Name, metav1.GetOptions{})
+					if err != nil {
+						return nil, err
+					}
+
+					return util.ParseNodeHostSubnetAnnotation(updatedNode, ovntypes.DefaultNetworkName)
+				}, 2).Should(gomega.Equal(ovntest.MustParseIPNets("10.128.0.0/24")))
+
+				return nil
+			}
+
+			err := app.Run([]string{
+				app.Name,
+			})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		})
+
+		ginkgo.It("allocates a subnet for a new node", func() {
```

Comment:
fyi, I didn't check but thinking this test must exist in clustermanager_test.go, I seem to remember numan adding it when he did the initial PR for a new node add event.. but we could remove that eventually now that we've started a new test file for ncc.